### ISSUE TITLE
Support for Inspecting/Reflecting PostgreSQL column defaults using GENERATED

### ISSUE THREAD

**Issue by sigmavirus24** [sqlalchemy/sqlalchemy#5324]
**Is your feature request related to a problem? Please describe.**

Today, I use SQLAlchemy to reflect and inspect Postgres tables. People are starting to move away from `SERIAL` types to using `GENERATED BY DEFAULT AS IDENTITY` and `GENERATED ALWAYS AS IDENTITY`. 

For example if I:

```sql
CREATE TABLE testidentity (id BIGINT PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY, name VARCHAR NOT NULL, optional VARCHAR);
```

When I use `sqlalchemy.inspect` to eventually `get_columns` I'm given:

```py
[{'name': 'id',
  'type': BIGINT(),
  'nullable': False,
  'default': None,
  'autoincrement': False,
  'comment': None},
 {'name': 'name',
  'type': VARCHAR(),
  'nullable': False,
  'default': None,
  'autoincrement': False,
  'comment': None},
 {'name': 'optional',
  'type': VARCHAR(),
  'nullable': True,
  'default': None,
  'autoincrement': False,
  'comment': None}]
```

But there is a default and it does autoincrement

**Describe the solution you'd like**

Support for describing these generated defaults as having a default and as autoincrementing (although the latter is less important)

**Describe alternatives you've considered**

Trying to inspect the metadata via some other library directly and working around SQLAlchemy

**Additional context**

N/A

**Have a nice day!**

**Comment by sigmavirus24** [comment]
I'm also willing to assist with implementing this but I have no clue as to where to start as I've never contributed to this project before.

**Comment by zzzeek** [comment]
this is possibly not trivial, not on the reflection side, but on the mechanics of INSERT.    all of the PG internals assume SERIAL for autoincrement and getting this format to work would be the hard part.

**Comment by jvanasco** [comment]
Further Reading:

* [PostgreSQL 10 identity columns explained](
https://www.2ndquadrant.com/en/blog/postgresql-10-identity-columns/) _Overview on Identity Columns by a PostgreSQL developer who (co?)authored the feature_
* [PostgreSQL Docs: Generated Columns](https://www.postgresql.org/docs/12/ddl-generated-columns.html)

**Comment by zzzeek** [comment]
we do support computed columns in general however i think PG's "identity" is a separate syntax.

**Comment by jvanasco** [comment]
Aside from reflection:

* new column type and DDL.  
* potential config options: `generated=OneOf(["always", "default"])`, `**sequence_options`
* inserts will need the potential for `OVERRIDING SYSTEM VALUE`

@zzzeek check out this link that goes into some of what is done regarding the auto-creation of serials.  https://www.depesz.com/2017/04/10/waiting-for-postgresql-10-identity-columns/

**Comment by CaselIT** [comment]
We should check if the `generated always as identity` is already piked up by the reflection logic in 1.4, since the syntax is the same as a normal compute column.

**Comment by zzzeek** [comment]
> We should check if the `generated always as identity` is already piked up by the reflection logic in 1.4, since the syntax is the same as a normal compute column.

the syntax is not what we generate, I tried this:
```

class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, Computed("AS IDENTITY"), primary_key=True)
    data = Column(String)
e = create_engine("postgresql://scott:tiger@localhost/test", echo='debug')

Base.metadata.drop_all(e)
Base.metadata.create_all(e)


```

and got

```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "("
LINE 3:  id INTEGER GENERATED ALWAYS AS (AS IDENTITY) STORED NOT NUL...
                                        ^

[SQL: 
CREATE TABLE a (
	id INTEGER GENERATED ALWAYS AS (AS IDENTITY) STORED NOT NULL, 
	data VARCHAR, 
	PRIMARY KEY (id)
)

```

I think this looks like a totally different (but very similar) syntax.    One proposal is to add this as a column option:
```

class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, postgresql_identity="BY DEFAULT", primary_key=True)
    data = Column(String)


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, postgresql_identity="ALWAYS", primary_key=True)
    data = Column(String)

```

**Comment by zzzeek** [comment]
well great news, as someone has created this amazing internal primary key architecture for the past 15 years, it just works as far as doing the RETURNING and all that.   so this should not be hard, here is the POC:
```

diff --git a/lib/sqlalchemy/dialects/postgresql/base.py b/lib/sqlalchemy/dialects/postgresql/base.py
index 08f1e018e..3cd8d0a23 100644
--- a/lib/sqlalchemy/dialects/postgresql/base.py
+++ b/lib/sqlalchemy/dialects/postgresql/base.py
@@ -1909,9 +1909,15 @@ class PGDDLCompiler(compiler.DDLCompiler):
         if isinstance(impl_type, sqltypes.TypeDecorator):
             impl_type = impl_type.impl
 
+        if column.primary_key:
+            identity = column.dialect_options['postgresql']['identity']
+        else:
+            identity = None
+
         if (
             column.primary_key
             and column is column.table._autoincrement_column
+            and identity is None
             and (
                 self.dialect.supports_smallserial
                 or not isinstance(impl_type, sqltypes.SmallInteger)
@@ -1936,9 +1942,19 @@ class PGDDLCompiler(compiler.DDLCompiler):
                 type_expression=column,
                 identifier_preparer=self.preparer,
             )
-            default = self.get_column_default_string(column)
-            if default is not None:
-                colspec += " DEFAULT " + default
+            if identity:
+                colspec += " GENERATED %s AS IDENTITY " % identity
+                # TODO: sequence options here, while I wanted to
+                # try postgresql_identity_start, postgresql_identity_increment,
+                # there are so many options at https://www.postgresql.org/docs/10/sql-createsequence.html
+                # it is easier just to let them type out the full set
+                # of keywords
+                # postgresql_identity_sequence_options="INCREMENT BY 5 MINVALUE 2 OWNED BY foo.bar"
+
+            else:
+                default = self.get_column_default_string(column)
+                if default is not None:
+                    colspec += " DEFAULT " + default
 
         if column.computed is not None:
             colspec += " " + self.process(column.computed)
@@ -2492,6 +2508,13 @@ class PGDialect(default.DefaultDialect):
                 "inherits": None,
             },
         ),
+        (
+            schema.Column,
+            {
+                "identity": None,
+                "identity_sequence_options": None,
+            },
+        ),
     ]
 
     reflection_options = ("postgresql_ignore_search_path",)

```

with that patch I can run this script and it uses RETURNING and gets the pk back:

```
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, postgresql_identity="BY DEFAULT", primary_key=True)
    data = Column(String)


e = create_engine("postgresql://scott:tiger@localhost/test", echo="debug")

Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

a1 = A(data="foo")

s.add(a1)
s.flush()


assert a1.id == 1


```

**Comment by CaselIT** [comment]
> try postgresql_identity_start, postgresql_identity_increment,
   there are so many options at https://www.postgresql.org/docs/10/sql-createsequence.html
   it is easier just to let them type out the full set
   of keywords
   postgresql_identity_sequence_options="INCREMENT BY 5 MINVALUE 2 OWNED BY foo.bar"

can't we just pass a `Sequence` object instead since we already have it implemented? `postgresql_identity_sequence= Sequence(start=42, increment=-3)`?

**Comment by CaselIT** [comment]
The only problem may be the name of the sequence, since I believe its currently required in `Sequence` while I believe it cannot be supplied when using identity columns

**Comment by zzzeek** [comment]
the whole thing with passing the sequence is exactly what I did with SQL Server, which I had to reverse in 6629d9f89273eda53a578cce41af6c79135254c7 when SQL Server had real sequences.     Passing a Sequence to Column means that you want that "CREATE SEQUENCE" to happen and I think we run into trouble when we repurpose things out of convenience.

it's also possible that here if a Sequence is present without the optional=True flag, and the postgresql_identity flags, that a warning or an error is raised as the intent is ambiguous in this case.     the optional=True flag indicates that the sequence / CREATE SEQUENCE should be used only if the backend database does not support any other kind of autoincrement, e.g. Oracle and Firebird.

**Comment by CaselIT** [comment]
Maybe we can subclass sequence in the pg dialect, creating a class called something like `class ItentityOptions(Sequence):` with a different visit name and a default `name` that is ignored, so that we would have all the options and documentation, but we are not passing a sequence.

But this may make reflection harder to implement, so I'm not sure if it's better than just adding the text.

**Comment by jvanasco** [comment]
Re: reflection

The infoschema columns on 10/11/12 are the same (https://www.postgresql.org/docs/12/infoschema-columns.html)

Name	Data Type	Description
is_identity	yes_or_no	If the column is an identity column, then YES, else NO.
identity_generation	character_data	If the column is an identity column, then ALWAYS or BY DEFAULT, reflecting the definition of the column.
identity_start	character_data	If the column is an identity column, then the start value of the internal sequence, else null.
identity_increment	character_data	If the column is an identity column, then the increment of the internal sequence, else null.
identity_maximum	character_data	If the column is an identity column, then the maximum value of the internal sequence, else null.
identity_minimum	character_data	If the column is an identity column, then the minimum value of the internal sequence, else null.
identity_cycle	yes_or_no	If the column is an identity column, then YES if the internal sequence cycles or NO if it does not; otherwise null.

**Comment by zzzeek** [comment]
if we make it the text for now, we have a clear forwards-upgrade path to use something nicer in the future.   on SQL Server we have indiviudual keys for each field.   I would want to address both dialects at once.

**Comment by jvanasco** [comment]
regarding @CaselIT's suggestion... what if the argument to the column was just `identity_sequence=Sequence(*args, **wargs)`.  The Postgres and SqlServer dialects could then implement/ignore the sequence as the identity as needed, but other dialects would ignore the sequence completely?

**Comment by CaselIT** [comment]
Another alternative is to use a function that takes the attributes and returns them a simple `tuple(args, kwargs)` similar to `bar_ref`

**Comment by sigmavirus24** [comment]
I'm a bit lost in all of this, so let me know when y'all have a solid plan and if you need help implementing it that I can help with.

**Comment by zzzeek** [comment]
@caselit - let's use IdentityOptions however lets switch it around and make it a mixin:

```
class IdentityOptions(object):
    # defines options for IDENTITY or SEQUENCE objects

class Sequence(IdentityOptions, SchemaItem):
    # a SEQUENCE
```

for this object we'd need to set up public API function + docstring + has to work for MSSQL also.    I think that is worthwhile feature addition but it is not necessarily a requirement to just get PG IDENTITY off the ground.

**Comment by zzzeek** [comment]
> I'm a bit lost in all of this, so let me know when y'all have a solid plan and if you need help implementing it that I can help with.

we're just haggling over how the various sequence options would be specified for this new IDENTITY construct.   the main task of implementing it remains straightforward starting with the POC I have above and then getting some test coverage, which in this case I think should be specific under test/dialect/postgresql/ , including test_compiler.py for DDL tests, test_reflection.py for testing that it is reflected, test_dialect.py for some round trip tests on creating a table, inserting rows, and getting primary key identities back.

**Comment by CaselIT** [comment]
Regarding `IdentityOptions`, it makes sense, so that it can be shared with mssql.

I agree that for this release issue we could support the `postgresql_identity_sequence_options` as string, and in the future we will also accept an instance of `IdentityOptions`.
The only thing I'm wondering if this complicates reflection since I don't think the db can return the sequence option sql string.

**Comment by CaselIT** [comment]
As an alternative we could get the `IdentityOptions ` going fist, just as a construct, then use it for this issue and then for mssql?

**Comment by zzzeek** [comment]
either way is fine but having it as a separate thing is probably nice

**Comment by CaselIT** [comment]
Ok, I'll get `IdentityOptions` going.

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Introduce :class:`.IdentityOptions` to store common parameters for** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1973

**Comment by CaselIT** [comment]
I've provided a first implementation of IdentityOptions.

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Introduce :class:`.IdentityOptions` to store common parameters for** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1973

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Introduce :class:`.IdentityOptions` to store common parameters for** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1979

**Comment by CaselIT** [comment]
@zzzeek there seem to be a bug in the `References` hook, since every time a new revision is added the bot makes a new comment. That does not happen in the `Fixes` hook

(sorry for the ot)

**Comment by zzzeek** [comment]
assuming that comes from the hook here: https://github.com/sqlalchemyorg/publishthing/blob/master/publishthing/apps/gerrit_patchset_comment.py   both fixes and references are treated the same way.  it's looking for new gerrit patchsets where the commit message has changed since the last one.

**Comment by CaselIT** [comment]
Probably best to continue on that repo (or not to continue at all :) )

Seems caused by me who fixed the commit message. Maybe instead of the "message", "subject" could be used for the comparison?
Still, it's not the first time I update a commit message and I've never noticed duplicated comments for the fixes hook, but they are indeed the same code, so 🤷‍♂️

For reference this is the message
```json
)]}'
{
  "commit": "e1d8b15b636793a84f0d2baa7d6213ce7dec8080",
  "parents": [
    {
      "commit": "79de84b25e87bbb7fa94f0dd513b4abc76e05a7e",
      "subject": "Actively unset reset agent in discard transaction"
    }
  ],
  "author": {
    "name": "Federico Caselli",
    "email": "<email here>",
    "date": "2020-05-14 20:57:56.000000000",
    "tz": 120
  },
  "committer": {
    "name": "Federico Caselli",
    "email": "<email here>",
    "date": "2020-05-15 17:11:04.000000000",
    "tz": 120
  },
  "subject": "Introduce :class:`.IdentityOptions` to store common parameters for sequences and identity columns.",
  "message": "Introduce :class:`.IdentityOptions` to store common parameters for\nsequences and identity columns.\n\nReferences: #5324\nChange-Id: I72f7fc1a003456206b004d3d26306940f9c36414\n"
}
```

**Comment by CaselIT** [comment]
Ok, just to recap.

Once https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1973 is merged we can use the `IdentityOptions` as `postgresql_identity_options=IdentityOptions(start=42, increment=-3)` so we can probably leverage most of the compiler logic used for the sequences to render the options.
This class can also be used when reflecting the columns from the db.

```py

class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, postgresql_identity="BY DEFAULT", primary_key=True, postgresql_identity_options=IdentityOptions(start=42, increment=-3))
    data = Column(String)


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, postgresql_identity="ALWAYS", primary_key=True, postgresql_identity_options=IdentityOptions(start=-20, increment=3))
    data = Column(String)
```

**Comment by zzzeek** [comment]
> Probably best to continue on that repo (or not to continue at all :) )


continuing the OT part of this regarding the "references" hook, it compares the whole text because it is looking to see if new issue numbers have appeared which need to be referenced.

However, this can be easily changed to do the right thing but diffing the set of issue numbers (And the type "fixes" or "references") that are referenced in the complete text, rather than the whole text itself.

that way the message would only go out when an issue number is newly referenced.   Or in the case where one is removed then added again later but that is not common.

**Comment by CaselIT** [comment]
@sigmavirus24 if you want to give it a go, `IdentityOptions` has been merged

**Comment by CaselIT** [comment]
I think we should probably continue in #5362 since this feature is not pg only it probably makes more sense to add support for it not a pg only add-on

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Add support for identity columns** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2018

**Comment by sigmavirus24** [comment]
Sorry, I lost track of this. Looking at https://docs.sqlalchemy.org/en/13/changelog/changelog_13.html this looks like it would land in 1.3.20 but doesn't have a changelog item yet, is that understanding accurate?

**Comment by CaselIT** [comment]
it was re-milestoned to 1.4

**Comment by CaselIT** [comment]
Actually now that I read this issue better, it should be reopened, since reflection is yet to implement.

Currently been tracked at https://github.com/sqlalchemy/sqlalchemy/issues/5527

**Comment by zzzeek** [comment]
is one of #5324 or #5527 a dupe of the other?

**Comment by jvanasco** [comment]
@CaselIT has been tracking this feature as such:

* #5324 - postgres specific support for identity
* #5360 - oracle specific support for identity
* #5362 - "generated as identity"
* #5527 - reflecting the identity columns

**Comment by zzzeek** [comment]
alrighty ill leave it to the experts

**Comment by CaselIT** [comment]
@zzzeek Basically yes, #5527 is more general than this issue, since it's for all backends. We can keep this open o close it pointing to the other one. But this will be closed together with #5527.

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Add reflection for Identity columns** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2248
